### PR TITLE
Bluetooth: Shell: Select BT_TICKER_NEXT_SLOT_GET when BT_LL_SW_SPLIT

### LINKED ISSUE #51631: bluetooth: shell: linker error
**Describe the bug**

While trying to build the Bluetooth shell, I faced the following linker error:

```
ction=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr && /usr/bin/cmake -E echo
/home/gmarull/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/subsys/bluetooth/shell/libsubsys__bluetooth__shell.a(ticker.c.obj): in function `cmd_ticker_info':
/home/gmarull/ws/nordic/zephyrproject/zephyr/subsys/bluetooth/shell/ticker.c:62: undefined reference to `ticker_next_slot_get'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

If the BT shell depends on another module being active (I suspect that this is the case), it should be specified in Kconfig either with a dependency or selection.

**To Reproduce**

```
west build -b nrf52840dk_nrf52840 samples/hello_world -p -- -DCONFIG_BT=y -DCONFIG_BT_SHELL=y
```

Note: I guess I'm missing an option, but it's not easy to know which one since `CONFIG_BT_SHELL` only selects `SHELL`.

**Expected behavior**

I can build the BT shell if `CONFIG_BT_SHELL=y`, or I get a Kconfig error/warning saying what is missing.

**Impact**

Annoyance

**Logs and console output**
See above

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.15.0
 - Commit SHA or Version used: 2627063ab21913bf82c5342ce115af1bf18e4299

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by cvinayak**
Select BT_TICKER_NEXT_SLOT_GET when BT_LL_SW_SPLIT enabled and building applications with shell support (BT_SHELL).

Fixes #51631.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

**Comment by cvinayak** [comment]
@gmarull this PR fixes the linker error at my end, do test and provide your approval.

**Comment by Thalley** [comment]
@cvinayak It honestly seems odd to `select BT_TICKER_NEXT_SLOT_GET` in the shell Kconfig. 

How about we simply guard the `ticker.c` shell commands by `CONFIG_BT_TICKER_NEXT_SLOT_GET`? The issues seems that `cmd_ticker_info` expects it to be enabled. 

If the `CONFIG_BT_TICKER_NEXT_SLOT_GET` is optional to enable, then IMO the shell command should be optionally included instead.

**Review by alwa-nordic — APPROVED** [review]

**Comment by cvinayak** [comment]
> @cvinayak It honestly seems odd to `select BT_TICKER_NEXT_SLOT_GET` in the shell Kconfig.
> 
> How about we simply guard the `ticker.c` shell commands by `CONFIG_BT_TICKER_NEXT_SLOT_GET`? The issues seems that `cmd_ticker_info` expects it to be enabled.
> 
> If the `CONFIG_BT_TICKER_NEXT_SLOT_GET` is optional to enable, then IMO the shell command should be optionally included instead.

@Thalley When using Zephyr Controller (`BT_LL_SW_SPLIT`) `cmd_ticker_info` provides information of the controller scheduling, which if is optional, then users of shell will not be aware of such a command unless explicitly adding `CONFIG_BT_TICKER_NEXT_SLOT_GET` in prj.conf. Default `CONFIG_BT=y` only enables broadcast role on nRF5x SoCs, which do not use `BT_TICKER_NEXT_SLOT_GET` but a shell can use it to provide information of the radio event scheduling.

**Comment by Thalley** [comment]
> then users of shell will not be aware of such a command unless explicitly adding CONFIG_BT_TICKER_NEXT_SLOT_GET in prj.conf. 

But is that bad? IMO if it is an optional feature, which it clearly is, then to me it only makes sense if the command is also optional. 

To me this is not any different from e.g. 
```c
#if defined(CONFIG_SETTINGS)
	SHELL_CMD_ARG(settings-load, NULL, HELP_NONE, cmd_settings_load, 1, 0),
#endif
```
where obviously we can only load the settings if we have enabled settings. Likewise we can only get the next ticker slot if `CONFIG_BT_TICKER_NEXT_SLOT_GET` is enabled. 

I won't block the PR as it is, as it does solve the raised issue, but I am also not going to approve it as is (at least not without more arguments as to why we don't use guard the command).

**Comment by alwa-nordic** [comment]
I unwillingly agree with @Thalley. The precedence is to hide commands that rely on features that are not compiled in.

Out of PR scope RFC: I think we should instead keep the commands around in the help text and give a helpful "feature not compiled-in" when invoked.

**Comment by cvinayak** [comment]
> To me this is not any different from e.g.
> 
> ```c
> #if defined(CONFIG_SETTINGS)
> 	SHELL_CMD_ARG(settings-load, NULL, HELP_NONE, cmd_settings_load, 1, 0),
> #endif
> ```
> 
> where obviously we can only load the settings if we have enabled settings. Likewise we can only get the next ticker slot if `CONFIG_BT_TICKER_NEXT_SLOT_GET` is enabled.

To me `cmd_ticker_info` is similar to `cmd_init`;  `cmd_init` (which uses `bt_enable()`) is provided by `CONFIG_BT_SHELL`, and hence it is necessary to select `CONFIG_BT`. `cmd_ticker_info` (which uses `ticker_next_slot_get()`) is provided by `CONFIG_BT_SHELL` when `CONFIG_BT_LL_SW_SPLIT=y` and hence it is necessary to select `CONFIG_BT_TICKER_NEXT_SLOT_GET`

**Comment by Thalley** [comment]
> > To me this is not any different from e.g.
> > ```c
> > #if defined(CONFIG_SETTINGS)
> > 	SHELL_CMD_ARG(settings-load, NULL, HELP_NONE, cmd_settings_load, 1, 0),
> > #endif
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > where obviously we can only load the settings if we have enabled settings. Likewise we can only get the next ticker slot if `CONFIG_BT_TICKER_NEXT_SLOT_GET` is enabled.
> 
> To me `cmd_ticker_info` is similar to `cmd_init`; `cmd_init` (which uses `bt_enable()`) is provided by `CONFIG_BT_SHELL`, and hence it is necessary to select `CONFIG_BT`. `cmd_ticker_info` (which uses `ticker_next_slot_get()`) is provided by `CONFIG_BT_SHELL` when `CONFIG_BT_LL_SW_SPLIT=y` and hence it is necessary to select `CONFIG_BT_TICKER_NEXT_SLOT_GET`

`cmd_ticker_info` is based on an optional feature, `cmd_init` is not, so I don't think that is a fair comparison.

**Comment by cvinayak** [comment]
> `cmd_ticker_info` is based on an optional feature, `cmd_init` is not, so I don't think that is a fair comparison.

Refer to #51631 

`west build -b nrf52840dk_nrf52840 samples/hello_world -p -- -DCONFIG_BT=y -DCONFIG_BT_SHELL=y`

`CONFIG_BT` and `CONFIG_BT_SHELL`  are optional feature for the `hello_world` sample.

**Comment by Thalley** [comment]
> > `cmd_ticker_info` is based on an optional feature, `cmd_init` is not, so I don't think that is a fair comparison.
> 
> Refer to #51631
> 
> `west build -b nrf52840dk_nrf52840 samples/hello_world -p -- -DCONFIG_BT=y -DCONFIG_BT_SHELL=y`
> 
> `cmd_init` provided by `CONFIG_BT` and `CONFIG_BT_SHELL` are optional feature for the `hello_world` sample.

Yes, and enabling them doesn't cause issues, because the commands are properly guarded. Any other optional BT feature which isn't added is likewise not compiled and linked. `cmd_ticker_info` is the outlier here, which doesn't behave like any of other optional commands we have.

**Comment by cvinayak** [comment]
> `cmd_ticker_info` is the outlier here, which doesn't behave like any of other optional commands we have.

`cmd_ticker_info` is a feature (not optional command when zephyr controller is used with shell) in `CONFIG_BT_SHELL` exactly like `cmd_init` hence the `hello_world` explicitly selects `CONFIG_BT=y` in order to use `cmd_init`.   The build will have `CONFIG_BT_SHELL=n` if the explicit `CONFIG_BT=y` is not supplied. This PR is saving the user from explicitly supplying a `-DCONFIG_BT_TICKER_NEXT_SLOT_GET=y`.

`CONFIG_BT` and `CONFIG_BT_SHELL`  have a single parent-child dependency. `CONFIG_BT_SHELL` is hence guarded and detected by Kconfig if its dependency is missing. But for `CONFIG_BT_TICKER_NEXT_SLOT_GET` which has multiple parents like `CONFIG_BT_SHELL`, `CONFIG_BT_CTLR_ADV_PERIODIC`, `CONFIG_BT_CTLR_ADVANCED_SCHEDULING` etc... the two option are:

1. to add `default y if BT_SHELL || ....` . But then the disadvantage is that user can set an explicit `n` and still get a linker error; will need conditional compiles inside `cmd_ticker_info` definitions and beats the purpose of using separate c files to add shell commands. Or change the conditional include in CMakelist file and make `cmd_ticker_info` optional.
2. to have `select BT_TICKER_NEXT_SLOT_GET if BT_SHELL` in the `Kconfig.ll_sw_split` file. But this is something not done in any of the Bluetooth Kconfig files.

**Comment by Thalley** [comment]
> `cmd_ticker_info` is a feature (not optional command when zephyr controller is used with shell)

I don't understand - Why is `cmd_ticker_info` not optional? Seems like something the shell works absolutely fine without. To my understanding, that command should be optional, and as such, should be guarded by the `BT_TICKER_NEXT_SLOT_GET`. 

If you disagree, can you elaborate on why you don't consider `cmd_ticker_info` and optional shell command?

**Comment by cvinayak** [comment]
> If you disagree, can you elaborate on why you don't consider `cmd_ticker_info` and optional shell command?

Because `cmd_ticker_info` and others like `cmd_ll_addr_read` (erroneously removed in b67a31e) where shell features provided when Zephyr Controller is included. `test_tx`, `test_rx` and `test_end` of DTM commands where optional hence they were guarded by Kconfig.

See,

https://github.com/zephyrproject-rtos/zephyr/blob/3a3cfbb652b29e9f39a38fe6febf29e1570d32bc/subsys/bluetooth/shell/CMakeLists.txt#L113-L120

**Comment by Thalley** [comment]
> > If you disagree, can you elaborate on why you don't consider `cmd_ticker_info` and optional shell command?
> 
> Because `cmd_ticker_info` and others like `cmd_ll_addr_read` (erroneously removed in [b67a31e](https://github.com/zephyrproject-rtos/zephyr/commit/b67a31e411a761fd826b65b7d5e8a2c644366b14)) where shell features provided when Zephyr Controller is included. `test_tx`, `test_rx` and `test_end` of DTM commands where optional hence they were guarded by Kconfig.
> 
> See,
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/3a3cfbb652b29e9f39a38fe6febf29e1570d32bc/subsys/bluetooth/shell/CMakeLists.txt#L113-L120

But why can't it be optional? To me it seems that the BT shell can work perfectly fine without the command, so I would consider it optional by that reasoning.

**Comment by cvinayak** [comment]
> But why can't it be optional? To me it seems that the BT shell can work perfectly fine without the command, so I would consider it optional by that reasoning.

On similar lines as why `cmd_info` and many other informational shell commands that are not optional, and BT shell can work perfectly.

**Comment by Thalley** [comment]
> > But why can't it be optional? To me it seems that the BT shell can work perfectly fine without the command, so I would consider it optional by that reasoning.
> 
> On similar lines as why `cmd_info` and many other informational shell commands that are not optional, and BT shell can work perfectly.

`cmd_info` is in fact optional, and guarded by `CONFIG_BT_CONN`, so it isn't available for e.g. a device that is only a GAP Broadcaster or Observer. My point is that the only mandatory command for the BT shell is `cmd_init`, and everything else is guarded. 

Granted that `cmd_ticker_info` is guarded by  `if(CONFIG_BT_CTLR AND CONFIG_BT_LL_SW_SPLIT)`, but the command also depends on an additional Kconfiguration (`BT_TICKER_NEXT_SLOT_GET`) which is also optional. 

I'd be happy if we could simply add `CONFIG_BT_TICKER_NEXT_SLOT_GET` to `if(CONFIG_BT_CTLR AND CONFIG_BT_LL_SW_SPLIT)` to guard the inclusion of `ticker.c`, so that we don't need to select an optional feature to provide an optional command in the BT shell.

**Comment by cvinayak** [comment]
> I'd be happy if we could simply add `CONFIG_BT_TICKER_NEXT_SLOT_GET` to `if(CONFIG_BT_CTLR AND CONFIG_BT_LL_SW_SPLIT)` to guard the inclusion of `ticker.c`, so that we don't need to select an optional feature to provide an optional command in the BT shell.

I disagree. If I would make `bt_conn_get_info` optional by having a Kconfig because my production application does not need it, I would still have it `select` when I want to have a shell in my debug application.

**Comment by cvinayak** [comment]
image
image

Adding `CONFIG_BT_SHELL=y` to `samples/bluetooth/beacon`  will give minimal commands from various subsystems in use in the application.

**Comment by Thalley** [comment]
> > I'd be happy if we could simply add `CONFIG_BT_TICKER_NEXT_SLOT_GET` to `if(CONFIG_BT_CTLR AND CONFIG_BT_LL_SW_SPLIT)` to guard the inclusion of `ticker.c`, so that we don't need to select an optional feature to provide an optional command in the BT shell.
> 
> If I would make `bt_conn_get_info` optional by having a Kconfig because my production application does not need it, I would still have it `select` when I want to have a shell in my debug application.

Agreed! But that is not the case here. If the user want to have optional feature X available in their application, they should enable it in the application. The Kconfig `BT_SHELL` is not an application. The way that the application should be compiled if it wants `cmd_ticker_info`, then it should do 

```
west build -b nrf52840dk_nrf52840 samples/hello_world -p -- -DCONFIG_BT=y -DCONFIG_BT_SHELL=y -DCONFIG_BT_TICKER_NEXT_SLOT_GET=y
```

I doesn't seem like we'll come to an agreement here, and as I mentioned I won't block this PR, but I also won't approve it as is.

**Comment by carlescufi** [comment]
@Thalley I just want to point out that the cmds themselves are conditionally included in the shell based on the enabling of the Zephyr LL:

```
if(CONFIG_BT_CTLR AND CONFIG_BT_LL_SW_SPLIT)
  zephyr_library_sources(
    ll.c
    ticker.c
    )
```

**Review by carlescufi — APPROVED** [review]

**Review by jori-nordic — APPROVED** [review]

**Comment by Thalley** [comment]
> @Thalley I just want to point out that the cmds themselves are [conditionally included](https://github.com/zephyrproject-rtos/zephyr/blob/d8d81c4a09040f9d6f64b3cef134b5290151e777/subsys/bluetooth/shell/CMakeLists.txt#L113-L117) in the shell based on the enabling of the Zephyr LL:
> 
> ```
> if(CONFIG_BT_CTLR AND CONFIG_BT_LL_SW_SPLIT)
>   zephyr_library_sources(
>     ll.c
>     ticker.c
>     )
> ```

Correct, but the command in `ticker.c` depends on `CONFIG_BT_TICKER_NEXT_SLOT_GET`, but is not guarded. So compiling `ticket.c` when `CONFIG_BT_TICKER_NEXT_SLOT_GET=n` causes a linking error, and my suggestion is to guard calls to `ticker_next_slot_get` with `CONFIG_BT_TICKER_NEXT_SLOT_GET` to avoid this, instead of implicitly setting `CONFIG_BT_TICKER_NEXT_SLOT_GET=y` when enabling `CONFIG_BT_SHELL`. 

But I can see that I am the minority, so I will yield :)

**Comment by cvinayak** [comment]
> The Kconfig BT_SHELL is not an application.

I agree, `BT_SHELL` is not an application, it is a feature any application can include bringing in the Bluetooth Stack and provides default shell commands. I want the `cmd_ticker_info` be present as the default when Zephyr Controller is used.

> So compiling `ticker.c` when `CONFIG_BT_TICKER_NEXT_SLOT_GET=n` causes a linking error,

With changes in this PR, there will be no error in building Bluetooth samples with explicitly added `CONFIG_BT_SHELL=y` which enables the default host and controller commands... see for example of hello_world sample, even with explicit `CONFIG_BT_TICKER_NEXT_SLOT_GET=n`, a warning will be provided that it is forced `y` because I want  `cmd_ticker_info` to be default for including the Zephyr Controller:

image

**Comment by Thalley** [comment]
> I want the `cmd_ticker_info` be present as the default when Zephyr Controller is used.

But that's not what you are doing with this PR: What you are doing is `select`ing Kconfig that only has value for the Zephyr controller, and by `select`ing it, you are not making it "the default" but rather enforcing it to be enabled, without the possibility of disabling it. 

I would be fine if it was just enabled by default, but that is not what you are suggesting in the PR (even though it sounds like that is what you want). 

You previously suggested enabling it by default for the BT_SHELL (https://github.com/zephyrproject-rtos/zephyr/pull/51740#issuecomment-1295270438), which I would have been OK with, but that would still require conditionally including the command (which you seem to be against). 

In any case, I am still the minority here, and it has enough approvers to get merged and fix the linker issue :)

**Comment by cvinayak** [comment]
> But that's not what you are doing with this PR: What you are doing is `select`ing Kconfig that only has value for the Zephyr controller, and by `select`ing it, you are not making it "the default" but rather enforcing it to be enabled, without the possibility of disabling it.

The linker issue is a regression introduced in commit https://github.com/zephyrproject-rtos/zephyr/commit/367004fd57f05aced23005aa4d4e21272b7e5add, the `cmd_ticker_info` has always been enforced to be enabled. Possibility to disabling will be a new feature or change request to the default thats been so since atleast Zephyr v1.14 releases.

> You previously suggested enabling it by default for the BT_SHELL (https://github.com/zephyrproject-rtos/zephyr/pull/51740#issuecomment-1295270438), which I would have been OK with, but that would still require conditionally including the command (which you seem to be against).

You are welcome to open an enhancement issue to make `cmd_ticker_info` conditionally included in shell. May be a `BT_CTLR_SHELL` for Controller commands.

**Pull request merged**
